W5500.device.nut

```diff
@@ -0,0 +1,2538 @@
+// CLASS: W5500.Driver
+// ==============================================================================
+
+class W5500.Driver {
+
+    // Chip can support 8 sockets
+    static TOTAL_SUPPORTED_SOCKETS = 8;
+    static MAX_TX_MEM_BUFFER = 16;
+    static MAX_RX_MEM_BUFFER = 16;
+
+    // CLASS VARIABLES
+    _interruptPin = null;
+    _spi = null;
+    _cs = null;
+    _resetPin = null;
+
+    _connections = null; // open connections
+    _availableSockets = null; // array of sockets available
+    _maxNoOfSockets = 0; // max number of sockets wiznet may use
+    _noOfSockets = 0; // number of sockets
+    _socketMemory = null; // amount of memeory each socket has
+
+    // ***************************************************************************
+    // Constructor
+    // Returns: null
+    // Parameters:
+    //      spi - configured spi, W5500 supports spi mode 0 or 3
+    //      cs(optional) - configured chip select pin
+    //      reset(optional) - configured reset pin
+    // ***************************************************************************
+    constructor(interruptPin, spi, cs = null, resetPin = null) {
+
+        _interruptPin = interruptPin;
+        _spi = spi;
+        _cs = cs;
+        _resetPin = resetPin;
+        _connections = {};
+        _availableSockets = [];
+
+        // Check the pins
+        if (resetPin) _resetPin.configure(DIGITAL_OUT, 1);
+
+        local imp005 = ("spi0" in hardware);
+        if (_cs != null) {
+            _cs.configure(DIGITAL_OUT, 1);
+        } else if (!imp005) {
+            throw "You must pass in a chip select pin."
+        }
+
+        // Reset the hardware
+        reset();
+
+    }
+
+
+    // ***************************************************************************
+    // reset, note this is blocking for 0.2s
+    // Returns: this
+    // Parameters:
+    //          sw(optional) - boolean if true forces a software reset
+    //          Note: datasheet for W5500 states that software reset is
+    //                unreliable - don't use
+    // **************************************************************************
+    function reset(sw = false) {
+
+        // Stop the interrupt pin from firing
+        _interruptPin.configure(DIGITAL_IN);
+
+        // Clear the connections and sockets
+        _connections = {};
+        _availableSockets = [];
+
+        // Force disconnect/close all ports
+        forceCloseAllSockets();
+
+        // Reset chip to default state, blocks for 0.2s
+        // Note: Datasheet states that W5500 software reset is not reliable, use hardware reset.
+        if (sw || _resetPin == null) {
+            setMode(W5500_SW_RESET);
+            imp.sleep(0.2);
+        } else {
+            _resetPin.write(0);
+            imp.sleep(0.01); // hold at least 500us after assert low
+            _resetPin.write(1);
+        }
+        imp.sleep(0.2); // wait at least 150ms before configuring
```

Comment:
With MTIII it doesn't connect either and requires at least 1 sec delay in this case.